src/pymor/algorithms/gram_schmidt.py

```diff
@@ -59,7 +59,7 @@ def gram_schmidt(A, product=None, return_R=False, atol=1e-13, rtol=1e-13, offset
         A = A.copy()
 
     # main loop
-    R = np.eye(len(A))
+    R = np.eye(len(A), dtype=A.dtype)
```

Comment:
@renefritze, @pmli, I just noticed an issue here. In general, `A.dtype` will only be the correct `dtype` for `R` when `product is None`. In particular, `product` might be a complex matrix, in which case `R` will have complex entries. Probably the only way around this would be to add `dtype` to `OperatorInterface` - which sounds quite difficult. Overall I think we should reconsider, if we really need an `dtype` attribute. I think the current code, while not quite beautiful, is acceptable as well.

Reply:
Yes, it seems adding `dtype` to operators would be necessary. It could still be useful for having control over all `dtypes`, but that is probably a separate issue.

Reply:
Yeah, but we have to decide now what we really want: 

The changes in this PR will break `gram_schmidt` for certain situations and the above will only work if we take the `dtype` of `product` into account, whatever that is. The main motivation for the `dtype` attribute was to make these changes possible.

Adding the `dtype` attribute to `VectorArrayInterface` adds a few (surprising) complications. For instance:
- What is the `dtype` of a `view` in case of a `ListVectorArray` where the individual vectors have different `dtype` and the `view` only contains vectors of the smaller `dtype`?
- What is the `dtype` of `BlockVectorArray.dofs`, when the requested dofs are all from a single block with smaller `dtype`.

Such questions can have clear answers, and it is possible to get the implementation right. All this requires work, however, so we should be sure there is enough benefit. If there is only real benefit, when `dtype` is also introduced for `Operators`, we need to take this into account and I fear that this will even more difficult to define (think about a nonlinear `Operator` which is not represented by a matrix that (or its inverse) is applied to `VectorArrays` of different `dtypes`).

Reply:
I agree more thought needs to go into this. I also agree that this change will not work without a `dtype` for `Operators`. Maybe a good topic for the next dev meeting?

Maybe another option for `gram_schmidt`, which doesn't use `dtype` or `np.promote_types`, would be forming `R` as a list of lists and letting `np.array(R)` decide about the correct `dtype`. But, I would agree if you say that it's also not quite beautiful or that it's not worth making such a change.